### ISSUE TITLE
Fine-tuning the fluorescence nuclei and cells model on extra annotated data

### ISSUE THREAD

**Issue by walkeraw2002** [instanseg/instanseg#27]
Hello,

I have been working with the fluorescence nuclei and cells model, and I would like to fine-tune it on some additional annotated data that I have.

I used the download_model function, which provided the instanseg.pt file. From my understanding, I can extract the weights from this file to create a .pth file to resume training. Is this the correct approach for fine-tuning the pretrained model, or is there a better way to go about this?

Additionally, the dataset I want to fine-tune on contains multi-channel membrane images with corresponding membrane masks, but no nuclear masks. Is it possible to fine-tune just the membrane segmentation portion of the model without needing nuclear annotations, or will the model require both membrane and nuclear channels/masks for effective training?

Thank you very much, and congratulations on the development of InstanSeg! The model performs great and is much faster compared to other segmentation models I’ve used.

**Comment by ThibautGoldsborough** [comment]
Hello,

Unfortunately no, I'm nearly sure there's no way of going from the instanseg.pt file back to the model_weights.pth. and we haven't shared the raw model_weights.pth (although we will do fat some point). Sorry!

Interesting question for training, the truth is, I don't know. But if it were me I'd retrain the model from scratch with both the CPDMI dataset and your own annotations. Training should take around 24h or less on a standard GPU and we explain in the notebook how to download CPDMI. (If not, you run the risk that the model forgets what nuclei look like.) To do this you can specify:

python train.py -data segmentation_dataset.pth -source "[CPDMI_2023, YOUR_DATASET]" -target NC -ci True

By the way, if you make your annotations public then we might be able to do the training for you and make the model available (referencing your contribution).

**Comment by walkeraw2002** [comment]
Hi Thibaut,

Thank you very much for your reply! I will follow your advice and work on training a model on the CPDMI dataset along with my annotations.

Unfortunately, I do not have the permission/authority to release the annotated dataset that I have, but thank you for offering to do the training for and make the model available.

**Comment by walkeraw2002** [comment]
> Hello,
> 
> Unfortunately no, I'm nearly sure there's no way of going from the instanseg.pt file back to the model_weights.pth. and we haven't shared the raw model_weights.pth (although we will do fat some point). Sorry!
> 
> Interesting question for training, the truth is, I don't know. But if it were me I'd retrain the model from scratch with both the CPDMI dataset and your own annotations. Training should take around 24h or less on a standard GPU and we explain in the notebook how to download CPDMI. (If not, you run the risk that the model forgets what nuclei look like.) To do this you can specify:
> 
> python train.py -data segmentation_dataset.pth -source "[CPDMI_2023, YOUR_DATASET]" -target NC -ci True
> 
> By the way, if you make your annotations public then we might be able to do the training for you and make the model available (referencing your contribution).

Hi Thibaut,

In the above reply, you recommended training a model from scratch on the CPDMI_2023 dataset as well as my own annotations. I appreciate your advice and have been working on training a model using this strategy.
 
I also would like to try and recreate the fluorescence_nuclei_and_cells pre-trained model that you have made available so I can fine-tune that on my annotations as well. Would you be able to share which specific datasets you used to train this model as well as the general approach you took for training (e.g., training on all datasets simultaneously or sequentially)? I understand if you aren't able to give me that information since you haven't released the model weights yet.

Thank you very much for your help!

Andrew

**Comment by walkeraw2002** [comment]
> Hi Andrew,
> 
> The command for training the public model was:
> 
> python train.py -data segmentation_dataset.pth -source "[CPDMI_2023]" --num_epochs 500 --experiment_str fluorescence_nuclei_and_cells -target NC --channel_invariant True --requested_pixel_size 0.5 --augmentation_type heavy
> 
> F1 score peaked at epoch 134 so you could consider reducing the number of epochs.

Hi Thibaut,

I really appreciate your continued support with my implementation of InstanSeg.

I ran the command you provided to recreate the public model, and the F1 score peaked at approximately 0.43. I was just wondering if you would be able to confirm whether that F1 score is similar to the expected value, or if it indicates that I am doing something wrong with the training.

Thanks again!

**Comment by ThibautGoldsborough** [comment]
This is f1_score_joint printed during training? yes this sounds correct.

**Comment by ThibautGoldsborough** [comment]
Apologies, I noticed a bug when training on CPDMI on nuclei and cells simultaneously (model outputs would be nan on first epoch). Did you run into this issue?

This is now fixed.

**Comment by walkeraw2002** [comment]
> Apologies, I noticed a bug when training on CPDMI on nuclei and cells simultaneously (model outputs would be nan on first epoch). Did you run into this issue?
> 
> This is now fixed.

Thank you for letting me know! I did not run into this issue, but I'll keep an eye out for it moving forward.
